include/drivers/bbram.h

```diff
@@ -0,0 +1,267 @@
+
+/**
+ * API template to check for V SBY power failure.
+ *
+ * @see z_impl_bbram_get_vsby
+ */
+typedef int (*bbram_api_get_vsby)(const struct device *dev);
+
+/**
+ * API template to reset the V SBY status.
+ *
+ * @see z_impl_bbram_reset_vsby
+ */
+typedef int (*bbram_api_reset_vsby)(const struct device *dev);
+
+/**
+ * API template to check for V CC1 power failure.
+ *
+ * @see z_impl_bbram_get_vcc1
+ */
+typedef int (*bbram_api_get_vcc1)(const struct device *dev);
+
+/**
+ * API template to reset V CC1 status.
+ *
+ * @see z_impl_bbram_reset_vcc1
+ */
+typedef int (*bbram_api_reset_vcc1)(const struct device *dev);
+
+/**
+ * API template to read from BBRAM.
+ *
+ * @see z_impl_bbram_read
+ */
+typedef int (*bbram_api_read)(const struct device *dev, int offset, int size,
+			      uint8_t *data);
+
+/**
+ * API template to write to BBRAM.
+ *
+ * @see z_impl_bbram_write
+ */
+typedef int (*bbram_api_write)(const struct device *dev, int offset, int size,
+			       uint8_t *data);
+
+__subsystem struct bbram_driver_api {
+	bbram_api_get_ibbr get_ibbr;
+	bbram_api_reset_ibbr reset_ibbr;
+	bbram_api_get_vsby get_vsby;
+	bbram_api_reset_vsby reset_vsby;
+	bbram_api_get_vcc1 get_vcc1;
+	bbram_api_reset_vcc1 reset_vcc1;
+	bbram_api_read read;
+	bbram_api_write write;
+};
+
+__syscall int bbram_get_ibbr(const struct device *dev);
+
+/**
+ * Check if "Invalid Battery-Backed RAM". This may occur as a result to low
+ * voltage at the VBAT pin.
+ *
+ * @param dev BBRAM device pointer.
+ * @return 0 if the Battery-Backed RAM data is valid.
+ */
+static inline int z_impl_bbram_get_ibbr(const struct device *dev)
+{
+	const struct bbram_driver_api *api =
+		(const struct bbram_driver_api *)dev->api;
+
+	if (!api->get_ibbr) {
+		return -ENOTSUP;
+	}
+
+	return api->get_ibbr(dev);
+}
+
+__syscall int bbram_reset_ibbr(const struct device *dev);
+
+/**
+ * Reset the IBBR status (calling bbram_ibbr will return 0 after this).
+ *
+ * @param dev BBRAM device pointer.
+ * @return 0 after reset is complete.
+ * @see z_impl_bbram_ibbr
+ */
+static inline int z_impl_bbram_reset_ibbr(const struct device *dev)
+{
+	const struct bbram_driver_api *api =
+		(const struct bbram_driver_api *)dev->api;
+
+	if (!api->reset_ibbr) {
+		return -ENOTSUP;
+	}
+
+	return api->reset_ibbr(dev);
+}
+
+__syscall int bbram_get_vsby(const struct device *dev);
+
+/**
+ * Check for V SBY power failure. This will return an error if the V SBY power
+ * domain is turned on after it was off.
+ *
+ * @param dev BBRAM device pointer.
+ * @return 0 if V SBY power domain is in normal operation.
+ */
+static inline int z_impl_bbram_get_vsby(const struct device *dev)
+{
+	const struct bbram_driver_api *api =
+		(const struct bbram_driver_api *)dev->api;
+
+	if (!api->get_vsby) {
+		return -ENOTSUP;
+	}
+
+	return api->get_vsby(dev);
+}
+
+__syscall int bbram_reset_vsby(const struct device *dev);
+
+/**
+ * Reset the V SBY status (calling bbram_vsby will return 0 after this).
```

Comment:
But what is the V SBY status? That needs to be documented.

This API should stand alone and be implementable by any vendor

Reply:
Same here, V_{SBY} seemed to be fairly common on datasheets for Volt standby. I added the more versbose name to the docs of `z_impl_bbram_get_vsby`.